leaderelection/leaderelection.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package leaderelection
+
+import (
+	"errors"
+	"os"
+	"strings"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+const ConfigMapNameEnv = "CONFIG_LEADERELECTION_NAME"
+
+var errEmptyLeaderElectionConfig = errors.New("empty leader election configuration")
+
+func NewConfigFromMap(data map[string]string) (*Config, error) {
+	config := defaultConfig()
+
+	if resourceLock, ok := data["resourceLock"]; ok {
+		config.ResourceLock = resourceLock
+	}
+
+	if leaseDurationStr, ok := data["leaseDuration"]; ok {
+		if leaseDuration, err := time.ParseDuration(leaseDurationStr); err == nil {
```

Comment:
unless I'm totes reading this wrong, if a user specifies an invalid value here, it will not take effect and they will have no visibility that what they wanted didn't happen?
Should we at the very least log this error?
Just worrying about the lack of visibility here.

Reply:
That's a good point; I think that in the presence of an invalid value, you should receive the default value and have a log message. WDYT

Reply:
Yes, it gets the default value now, my worry was the lack of feedback, which the log message would provide.

Reply:
Three choices here:

1. If you supply an invalid value, it fails loudly (i.e. `logger.Fatal`) This is most appropriate for startup conditions, where it might block something like a rolling update.

2. If an invalid config is supplied, the entire config should be considered invalid and rejected. In this case, this would be adding an else to `return nil, err`

3. If a partially-invalid config is supplied, do the best you can, and log errors about the parts that didn't work.

I'd most prefer option 1, but given that we're driving this from dynamically-updated configmaps, I think option 2 would be better than option 3, because it makes the application of the ConfigMap much more deterministic (all or nothing, vs all-except-my-typos).

Reply:
So...  @nimakaviani added a validating webhook that allows us to register configmaps for validation and reject them outright if it returns an `error`.  So the right answer here is multi-part:
1. This should return an error
2. Each downstream webhook should add this to the list of configmaps to validate.

Then the configmap update will be outright rejected and never saved to etcd.